ambry-store/src/main/java/com.github.ambry.store/ScanResults.java

```diff
@@ -0,0 +1,194 @@
+    return requestedSegmentMap;
+  }
+
+  private Long generateBucketTime(long referenceTime) {
+    long remainder = referenceTime % bucketTimeSpan;
+    return remainder == 0 ? referenceTime : referenceTime - remainder + bucketTimeSpan;
+  }
+
+  /**
+   * Helper function to extract the valid data size of each segment and put them into a {@link SortedMap}.
+   * @param bucketEndTime all buckets prior to and including this bucketEndTime is aggregated as the delta to be applied
+   *                      on the base value to compute the valid data size
+   * @return mapping of segment names to their corresponding valid data size
+   */
+  SortedMap<String, Long> getValidDataSizePerSegment(Long bucketEndTime) {
+    SortedMap<String, Long> retMap = new TreeMap<>();
+    if (!segmentBuckets.isEmpty() && bucketEndTime >= segmentBuckets.firstKey()) {
+      SortedMap<Long, HashMap<String, Long>> bucketsInRange =
+          segmentBuckets.subMap(segmentBuckets.firstKey(), true, bucketEndTime, true);
+      for (Map.Entry<String, Long> entry : segmentBaseMap.entrySet()) {
+        Long value = entry.getValue();
+        for (HashMap<String, Long> segmentDelta : bucketsInRange.values()) {
+          if (segmentDelta.containsKey(entry.getKey())) {
+            value += segmentDelta.get(entry.getKey());
+          }
+        }
+        retMap.put(entry.getKey(), value);
+      }
+    } else {
+      retMap.putAll(segmentBaseMap);
+    }
+    return retMap;
+  }
+
+  /**
+   * Helper function to extract the valid data size of each container and put then into a nested {@link HashMap}.
+   * @param bucketEndTime all buckets prior to and including this bucketEndTime is aggregated as the delta to be applied
+   *                      on the base value to compute the valid data size
+   * @return nested {@link HashMap} of serviceIds to containerIds to their corresponding valid data size
+   */
+  HashMap<String, HashMap<String, Long>> getValidDataSizePerContainer(Long bucketEndTime) {
+    HashMap<String, HashMap<String, Long>> retMap = new HashMap<>();
+    if (!containerBuckets.isEmpty() && bucketEndTime >= containerBuckets.firstKey()) {
+      SortedMap<Long, HashMap<String, HashMap<String, Long>>> bucketsInRange =
+          containerBuckets.subMap(containerBuckets.firstKey(), true, bucketEndTime, true);
+      for (Map.Entry<String, HashMap<String, Long>> entry : containerBaseMap.entrySet()) {
+        HashMap<String, Long> containerMap = new HashMap<>();
+        for (Map.Entry<String, Long> innerEntry : entry.getValue().entrySet()) {
+          Long value = innerEntry.getValue();
+          for (HashMap<String, HashMap<String, Long>> containerDelta : bucketsInRange.values()) {
+            if (containerDelta.containsKey(entry.getKey()) &&
+                containerDelta.get(entry.getKey()).containsKey(innerEntry.getKey())) {
+              value += containerDelta.get(entry.getKey()).get(innerEntry.getKey());
```

Comment:
With the existing implementation, each bucket only stores the future delta within their own start time and end time. That is, not aggregated delta.